README.md
[![Build, Test & Publish Main](https://github.com/bbeesley/aws-blue-green-toolkit/actions/workflows/build-test-on-push.yml/badge.svg)](https://github.com/bbeesley/aws-blue-green-toolkit/actions/workflows/build-test-on-push.yml) [![codecov](https://codecov.io/gh/bbeesley/aws-blue-green-toolkit/branch/main/graph/badge.svg)](https://codecov.io/gh/bbeesley/aws-blue-green-toolkit) [![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/) [![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

# aws-blue-green-toolkit

Utility functions to help with performing blue/green and canary deployments in AWS infrastructure

## background

AWS support for blue/green and canary deployments is provided by CodeDeploy. CodeDeploy is great in what it does, but it only really manages swapping versions for web services. Any services that sit behind the public facing API don't get managed by the CodeDeploy flow.

A common pattern for blue/green and canary deployments is to not just use a pair of services for the web component, but use a pair of services for the whole pipeline, eg data ingestion, databases, queues, data manipulation lambdas etc. You can manage starting, hydrating, and using these services as part of the CodeDeploy pipeline through the use of hooks (in the form of lambdas). Using these hook lambdas you can manage blue/green or canary deployments of a whole stack of services using the aws-sdk, unfortunately this normally means writing a lot of code to manage the process.

This module is a set of utility functions that I use to reduce the boilerplate required to set up these deploment hook lambdas.

## example architecture

For an example web api, using a node application in docker, with a SQL data store, and data updates being published on an SNS topic, the architecture may looks something like this:
![](docs/SimplifiedArchitecture.png)

In order to manage blue/green deployments for this service, CodeDeploy will directly handle deploying the new ECS service, attaching target groups to load balancers, flipping traffic over to the new service, and tearing down the old containers.

Unfortunately this really only covers less than half of the architecture. Everything from the data updates topic through to the database is up to you to manage through hooks. When deploying a new version of the service, the steps required will be something like this:

1.  Replacement database cluster is started or created
2.  Autoscaling minimum capacity on the new db cluster is set to match the current capacity of the active cluster
3.  Replacement database is purged and the latest schema is created
4.  SNS subscriptions are enabled for the replacement queue
5.  SQS subscription lambda is enabled for the replacement stack
6.  A new task set is created in the ECS service, it is bound to the replacement stack's database
7.  The new task set is placed in a testing target group, this is attached to a testing port on the load balancer
8.  Automated tests are carried out against the replacement service via the testing port
9.  The replacement stack becomes the active stack, what was the active stack is now the old stack.
10. Autoscaling minimum capacity on the active set is reverted to the normal value, database will scale in when traffic is lower.
11. SNS subscriptions are disabled for the old stack
12. SQS subscription for the old stack's ingest queue is disabled
13. Old database cluster is stopped or deleted
14. The old SQS queues are purged
15. The old ECS taskset is destroyed

CodeDeploy will carry out steps 6, 7, 9, and 15. This module contains tools to help you perform the other steps from your deployment hook lambda functions.

## api

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

*   [ClusterState](#clusterstate)
*   [EcsTools](#ecstools)
    *   [disableService](#disableservice)
        *   [Parameters](#parameters)
    *   [enableService](#enableservice)
        *   [Parameters](#parameters-1)
*   [SqsTools](#sqstools)
    *   [purgeQueues](#purgequeues)
        *   [Parameters](#parameters-2)
*   [DynamoTools](#dynamotools)
    *   [deleteTable](#deletetable)
        *   [Parameters](#parameters-3)
*   [CloudWatchTools](#cloudwatchtools)
    *   [disableAlarmsActions](#disablealarmsactions)
        *   [Parameters](#parameters-4)
    *   [enableAlarmsActions](#enablealarmsactions)
        *   [Parameters](#parameters-5)
*   [KinesisTools](#kinesistools)
    *   [deregisterConsumer](#deregisterconsumer)
        *   [Parameters](#parameters-6)
    *   [describeConsumer](#describeconsumer)
        *   [Parameters](#parameters-7)
    *   [registerConsumer](#registerconsumer)
        *   [Parameters](#parameters-8)
*   [StackReference](#stackreference)
*   [AuroraTools](#auroratools)
    *   [getClusterState](#getclusterstate)
        *   [Parameters](#parameters-9)
    *   [scaleIn](#scalein)
        *   [Parameters](#parameters-10)
    *   [scaleOut](#scaleout)
        *   [Parameters](#parameters-11)
    *   [getReaderCount](#getreadercount)
        *   [Parameters](#parameters-12)
    *   [startDatabase](#startdatabase)
        *   [Parameters](#parameters-13)
    *   [stopDatabase](#stopdatabase)
        *   [Parameters](#parameters-14)
    *   [deleteDatabase](#deletedatabase)
        *   [Parameters](#parameters-15)
    *   [applyTags](#applytags)
        *   [Parameters](#parameters-16)
    *   [enablePerformanceInsights](#enableperformanceinsights)
        *   [Parameters](#parameters-17)
*   [LambdaTools](#lambdatools)
    *   [createEventSourceMapping](#createeventsourcemapping)
        *   [Parameters](#parameters-18)
    *   [deleteEventMapping](#deleteeventmapping)
        *   [Parameters](#parameters-19)
    *   [disableEventMapping](#disableeventmapping)
        *   [Parameters](#parameters-20)
    *   [disableRule](#disablerule)
        *   [Parameters](#parameters-21)
    *   [enableEventMapping](#enableeventmapping)
        *   [Parameters](#parameters-22)
    *   [enableRule](#enablerule)
        *   [Parameters](#parameters-23)
    *   [getAlias](#getalias)
        *   [Parameters](#parameters-24)
    *   [getLatestMetrics](#getlatestmetrics)
        *   [Parameters](#parameters-25)
    *   [getVersion](#getversion)
        *   [Parameters](#parameters-26)
    *   [listEventSourceMappings](#listeventsourcemappings)
        *   [Parameters](#parameters-27)
*   [SnsTools](#snstools)
    *   [disableSubscription](#disablesubscription)
        *   [Parameters](#parameters-28)
    *   [enableSubscription](#enablesubscription)
        *   [Parameters](#parameters-29)

### ClusterState

[src/main/constants.ts:7-12](https://github.com/bbeesley/aws-blue-green-toolkit/blob/edb093206e6e38d2c6f0ad41d6f585adaf150024/src/main/constants.ts#L7-L12 "Source code on GitHub")

Enum for describing the state of an RDS cluster

Type: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)

### EcsTools

[src/main/ecs-tools.ts:11-64](https://github.com/bbeesley/aws-blue-green-toolkit/blob/edb093206e6e38d2c6f0ad41d6f585adaf150024/src/main/ecs-tools.ts#L11-L64 "Source code on GitHub")

Tools for managing pairs of ECS services

#### disableService

[src/main/ecs-tools.ts:32-43](https://github.com/bbeesley/aws-blue-green-toolkit/blob/edb093206e6e38d2c6f0ad41d6f585adaf150024/src/main/ecs-tools.ts#L32-L43 "Source code on GitHub")

Disables an ECS service by setting the desired task count to zero

##### Parameters

*   `reference` **[StackReference](#stackreference)** The stack to modify

Returns **any** {Promise<void>}

#### enableService

[src/main/ecs-tools.ts:52-63](https://github.com/bbeesley/aws-blue-green-toolkit/blob/edb093206e6e38d2c6f0ad41d6f585adaf150024/src/main/ecs-tools.ts#L52-L63 "Source code on GitHub")

Enables an ECS service by setting the desired task count to it's normal value

##### Parameters

*   `reference` **[StackReference](#stackreference)** The stack to modify

Returns **any** {Promise<void>}

### SqsTools

[src/main/sqs-tools.ts:11-50](https://github.com/bbeesley/aws-blue-green-toolkit/blob/edb093206e6e38d2c6f0ad41d6f585adaf150024/src/main/sqs-tools.ts#L11-L50 "Source code on GitHub")

Toolkit for SQS operations

#### purgeQueues

[src/main/sqs-tools.ts:32-45](https://github.com/bbeesley/aws-blue-green-toolkit/blob/edb093206e6e38d2c6f0ad41d6f585adaf150024/src/main/sqs-tools.ts#L32-L45 "Source code on GitHub")

Purges a queue pair (q and dlq) based on config and queue reference

##### Parameters

*   `reference` **[StackReference](#stackreference)** Reference to a subscription queue stack

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)\<void>**&#x20;

### DynamoTools

[src/main/dynamo-tools.ts:14-55](https://github.com/bbeesley/aws-blue-green-toolkit/blob/edb093206e6e38d2c6f0ad41d6f585adaf150024/src/main/dynamo-tools.ts#L14-L55 "Source code on GitHub")

Toolkit for Dynamo operations

#### deleteTable

[src/main/dynamo-tools.ts:35-48](https://github.com/bbeesley/aws-blue-green-toolkit/blob/edb093206e6e38d2c6f0ad41d6f585adaf150024/src/main/dynamo-tools.ts#L35-L48 "Source code on GitHub")

Deletes a dynamo table

##### Parameters

*   `reference` **[StackReference](#stackreference)** Reference to a active table

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)\<void>**&#x20;

### CloudWatchTools


CHANGELOG.md
## [3.6.3](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.6.2...v3.6.3) (2023-10-02)


### Bug Fixes

* no need to check specific error code text ([edb0932](https://github.com/bbeesley/aws-blue-green-toolkit/commit/edb093206e6e38d2c6f0ad41d6f585adaf150024))

## [3.6.2](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.6.1...v3.6.2) (2023-10-02)


### Bug Fixes

* dont check error type ([383460c](https://github.com/bbeesley/aws-blue-green-toolkit/commit/383460c14c621f9f2b9386a003981480033908bb))

## [3.6.1](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.6.0...v3.6.1) (2023-09-29)


### Bug Fixes

* use new InvalidDBInstanceStateFault class ([17da99d](https://github.com/bbeesley/aws-blue-green-toolkit/commit/17da99d964b79a2e17e5ae19b72033b8894af4da))

# [3.6.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.5.1...v3.6.0) (2023-09-28)


### Features

* **aurora-tools:** look up tags on cluster if none supplied in config ([64a594c](https://github.com/bbeesley/aws-blue-green-toolkit/commit/64a594c03eb36205f36bed5645411ef78bdade85))

## [3.5.1](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.5.0...v3.5.1) (2022-11-21)


### Bug Fixes

* send cluster data ([b9bf535](https://github.com/bbeesley/aws-blue-green-toolkit/commit/b9bf535db2a61efd28f12d0a7ee545f8860d8348))

# [3.5.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.4.1...v3.5.0) (2022-11-18)


### Features

* adds some initial tools for managing ECS service pairs ([42ba50a](https://github.com/bbeesley/aws-blue-green-toolkit/commit/42ba50a1d8991ce8192cea5e3ca17aec85dfce8c))

## [3.4.1](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.4.0...v3.4.1) (2022-10-14)


### Bug Fixes

* ensure legacy cjs output can be loaded ([5bbf9e6](https://github.com/bbeesley/aws-blue-green-toolkit/commit/5bbf9e641a10f7610cd3004ce462155b4ac23d05))

# [3.4.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.3.0...v3.4.0) (2022-09-29)


### Features

* add a retry with delay when db not ready ([26b71a7](https://github.com/bbeesley/aws-blue-green-toolkit/commit/26b71a7f90b4dd67749c0adc24342ab5aaade20d))

# [3.3.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.2.0...v3.3.0) (2022-09-02)


### Features

* allow re enabling insights if disabled ([72458a4](https://github.com/bbeesley/aws-blue-green-toolkit/commit/72458a411e0c376550d2f297da44f30d130b4e88))

# [3.2.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.1.0...v3.2.0) (2022-09-02)


### Features

* add back legacy support for cjs ([68f7603](https://github.com/bbeesley/aws-blue-green-toolkit/commit/68f76039aabfed1869f95f37e874e5b536b10869))

# [3.1.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.0.1...v3.1.0) (2022-08-26)


### Features

* adds enablePerformanceInsights method to AuroraTools ([4bf202e](https://github.com/bbeesley/aws-blue-green-toolkit/commit/4bf202e7d0ab69272e94719bf8a98029b57c725d))

## [3.0.1](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v3.0.0...v3.0.1) (2022-08-12)


### Bug Fixes

* only require namespace for aurora utils ([c1533b9](https://github.com/bbeesley/aws-blue-green-toolkit/commit/c1533b95ad7f7788a7630ff0002fc28351b63807))

# [3.0.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v2.1.1...v3.0.0) (2022-08-12)


### Bug Fixes

* update exports ([42d95dd](https://github.com/bbeesley/aws-blue-green-toolkit/commit/42d95dd7a33a43a8d8865b7e3d219b8714d3b151))


### Features

* migrate to aws sdk v3 and es modules ([dbd051a](https://github.com/bbeesley/aws-blue-green-toolkit/commit/dbd051ad67db544aadf1a642b0755cdceff988fa))


### BREAKING CHANGES

* Common JS is no longer supported

## [2.1.1](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v2.1.0...v2.1.1) (2021-07-08)


### Bug Fixes

* **build:** build es modules that support node 12 ([27251cf](https://github.com/bbeesley/aws-blue-green-toolkit/commit/27251cf00642449ae236d7b50de52ed4006bd31b))

# [2.1.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v2.0.1...v2.1.0) (2021-07-02)


### Features

* migrate to ES module exports ([15f68a4](https://github.com/bbeesley/aws-blue-green-toolkit/commit/15f68a42cbaee116a880269871f5acdf828fcac9))

## [2.0.1](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v2.0.0...v2.0.1) (2021-06-14)

# [2.0.0](https://github.com/bbeesley/aws-blue-green-toolkit/compare/v1.12.0...v2.0.0) (2021-06-03)


### Features

* **lambda-tools:** add a helper method to retrieve the latest lambda metrics ([2625bb8](https://github.com/bbeesley/aws-blue-green-toolkit/commit/2625bb8019f3c1ec6edb4404784091e862d486c7))
* drop support for node 10 ([988d4e2](https://github.com/bbeesley/aws-blue-green-toolkit/commit/988d4e2e962d11651fe5e50e4cb8287ddcfe00ea))


### BREAKING CHANGES


__mocks__/aws-sdk.js
const _rdsResponses = {
  describeDBClusters: {
    DBClusters: [
      {
        AllocatedStorage: 1,
        AvailabilityZones: ['eu-central-1a', 'eu-central-1b', 'eu-central-1c'],
        BackupRetentionPeriod: 5,
        DatabaseName: 'fn',
        DBClusterIdentifier: 'fn-db-b-dev',
        DBClusterParameterGroup: 'fn-db-b-dev',
        DBSubnetGroup: 'fn-db-b-dev',
        Status: 'stopped',
        EarliestRestorableTime: '2019-12-19T20:18:31.111Z',
        Endpoint:
          'fn-db-b-dev.cluster-0f0b5ec26c5f.eu-central-1.rds.amazonaws.com',
        ReaderEndpoint:
          'fn-db-b-dev.cluster-ro-0f0b5ec26c5f.eu-central-1.rds.amazonaws.com',
        CustomEndpoints: [],
        MultiAZ: true,
        Engine: 'aurora-postgresql',
        EngineVersion: '10.7',
        LatestRestorableTime: '2019-12-24T20:50:42.030Z',
        Port: 5432,
        MasterUsername: 'postgres',
        DBClusterOptionGroupMemberships: [],
        PreferredBackupWindow: '07:00-09:00',
        PreferredMaintenanceWindow: 'wed:03:00-wed:04:00',
        ReadReplicaIdentifiers: [],
        DBClusterMembers: [
          {
            DBInstanceIdentifier:
              'application-autoscaling-ddfc94a1-29af-4d17-b497-0f0b5ec26c5f',
            IsClusterWriter: false,
            DBClusterParameterGroupStatus: 'in-sync',
            PromotionTier: 15,
          },
          {
            DBInstanceIdentifier:
              'application-autoscaling-90ba4e59-e2df-488c-bbab-73422e4edea7',
            IsClusterWriter: false,
            DBClusterParameterGroupStatus: 'in-sync',
            PromotionTier: 15,
          },
          {
            DBInstanceIdentifier: 'fn-db-b-dev-1',
            IsClusterWriter: true,
            DBClusterParameterGroupStatus: 'in-sync',
            PromotionTier: 0,
          },
        ],
        VpcSecurityGroups: [
          { VpcSecurityGroupId: 'sg-bd3274a02951', Status: 'active' },
        ],
        HostedZoneId: 'FOOBARBAZBATFF',
        StorageEncrypted: false,
        DbClusterResourceId: 'cluster-FOOBARBAZBATFF',
        DBClusterArn: 'arn:aws:rds:eu-central-1:555:cluster:fn-db-b-dev',
        AssociatedRoles: [],
        IAMDatabaseAuthenticationEnabled: false,
        ClusterCreateTime: '2019-12-03T09:55:26.380Z',
        EnabledCloudwatchLogsExports: [],
        EngineMode: 'provisioned',
        DeletionProtection: false,
        HttpEndpointEnabled: false,
        ActivityStreamStatus: 'stopped',
        CopyTagsToSnapshot: false,
        CrossAccountClone: false,
      },
    ],
  },
};

const _rdsConstructor = jest.fn();
const _describeDBClusters = jest.fn(
  async () => _rdsResponses.describeDBClusters
);

class RDS {
  constructor(params) {
    _rdsConstructor(params);
  }
  describeDBClusters(params) {
    return { promise: _describeDBClusters.bind(this, params) };
  }
}

RDS._rdsConstructor = _rdsConstructor;
RDS._describeDBClusters = _describeDBClusters;
RDS._rdsResponses = _rdsResponses;

const _dynamoResponses = {
  deleteTable: {},
  waitFor: {},
};
const _dynamoConstructor = jest.fn();
const _deleteTable = jest.fn(async () => _dynamoResponses.deleteTable);
const _waitFor = jest.fn(async () => _dynamoResponses.waitFor);

class DynamoDB {
  constructor(params) {
    _dynamoConstructor(params);
  }
  deleteTable(params) {
    return { promise: _deleteTable.bind(this, params) };
  }
  waitFor(state, params) {
    return { promise: _waitFor.bind(this, state, params) };
  }
}

DynamoDB._dynamoConstructor = _dynamoConstructor;
DynamoDB._deleteTable = _deleteTable;
DynamoDB._waitFor = _waitFor;
DynamoDB._dynamoResponses = _dynamoResponses;

const _aasResponses = {
  describeScalableTargets: {
    ScalableTargets: [
      {
        ServiceNamespace: 'rds',
        ResourceId: 'cluster:fn-db-b-dev',
        ScalableDimension: 'rds:cluster:ReadReplicaCount',
        MinCapacity: 1,
        MaxCapacity: 15,
        RoleARN:
          'arn:aws:iam::555:role/aws-service-role/rds.application-autoscaling.amazonaws.com/AWSServiceRoleForApplicationAutoScaling_RDSCluster',
        CreationTime: '2019-12-03T09:56:17.243Z',
        SuspendedState: {
          DynamicScalingInSuspended: false,
          DynamicScalingOutSuspended: false,
          ScheduledScalingSuspended: false,
        },
      },
    ],
  },
  registerScalableTarget: {},
};
const _aasConstructor = jest.fn();
const _describeScalableTargets = jest.fn(
  async () => _aasResponses.describeScalableTargets
);
const _registerScalableTarget = jest.fn(
  async () => _aasResponses.registerScalableTarget
);

class ApplicationAutoScaling {
  constructor(params) {
    _aasConstructor(params);
  }
  describeScalableTargets(params) {
    return { promise: _describeScalableTargets.bind(this, params) };
  }
  registerScalableTarget(params) {
    return { promise: _registerScalableTarget.bind(this, params) };
  }
}

ApplicationAutoScaling._aasConstructor = _aasConstructor;
ApplicationAutoScaling._describeScalableTargets = _describeScalableTargets;
ApplicationAutoScaling._registerScalableTarget = _registerScalableTarget;
ApplicationAutoScaling._aasResponses = _aasResponses;

const _snsResponses = {
  setSubscriptionAttributes: {},
};
const _snsConstructor = jest.fn();
const _setSubscriptionAttributes = jest.fn(
  async () => _snsResponses.setSubscriptionAttributes
);

class SNS {
  constructor(params) {
    _snsConstructor(params);
  }
  setSubscriptionAttributes(params) {
    return { promise: _setSubscriptionAttributes.bind(this, params) };
  }
}

SNS._snsConstructor = _snsConstructor;
SNS._setSubscriptionAttributes = _setSubscriptionAttributes;
SNS._snsResponses = _snsResponses;
const _sqsResponses = {
  purgeQueue: {},
};
const _sqsConstructor = jest.fn();
const _purgeQueue = jest.fn(async () => _sqsResponses.purgeQueue);

class SQS {
  constructor(params) {
    _sqsConstructor(params);
  }
  purgeQueue(params) {
    return { promise: _purgeQueue.bind(this, params) };
  }
}

SQS._sqsConstructor = _sqsConstructor;
SQS._purgeQueue = _purgeQueue;
SQS._sqsResponses = _sqsResponses;
const _lambdaResponses = {
  updateEventSourceMapping: {},
  listEventSourceMappings: {
    NextMarker: null,
    EventSourceMappings: [
      {
        UUID: 'uuid',
        BatchSize: 10,
        MaximumBatchingWindowInSeconds: null,
        ParallelizationFactor: null,
        EventSourceArn: 'arn:aws:sqs:eu-central-1:555:fn-b-queue',
        FunctionArn: 'arn:aws:lambda:eu-central-1:555:function:fn-b',
        LastModified: new Date(),
        LastProcessingResult: null,
        State: 'Disabled',
        StateTransitionReason: 'USER_INITIATED',
        MaximumRecordAgeInSeconds: null,
        BisectBatchOnFunctionError: null,
        MaximumRetryAttempts: null,
      },
    ],
  },
  createEventSourceMapping: {
    UUID: 'e2fc5e11-79cc-47b1-bd80-fcd7a852b214',
    StartingPosition: 'TRIM_HORIZON',
    BatchSize: 100,
    MaximumBatchingWindowInSeconds: 0,
    ParallelizationFactor: 1,
    EventSourceArn:
      'arn:aws:kinesis:eu-central-1:123456789:stream/my-stream-dev',
    FunctionArn: 'arn:aws:lambda:eu-central-1:123456789:function:my-fn-dev',
    LastModified: '2020-12-09T11:26:31.934000+00:00',
    LastProcessingResult: 'No records processed',
    State: 'Creating',
    StateTransitionReason: 'User action',
    DestinationConfig: {
      OnFailure: {},
    },
    MaximumRecordAgeInSeconds: -1,
    BisectBatchOnFunctionError: false,
    MaximumRetryAttempts: -1,
  },
  deleteEventSourceMapping: {
    UUID: '091a66ea-1d4c-411c-97f0-039905401602',
    StartingPosition: 'TRIM_HORIZON',
    BatchSize: 100,
    MaximumBatchingWindowInSeconds: 0,
    ParallelizationFactor: 1,
    EventSourceArn:
      'arn:aws:kinesis:eu-central-1:123456789:stream/my-stream-dev',
    FunctionArn: 'arn:aws:lambda:eu-central-1:123456789:function:my-fn-dev',
    LastModified: '2020-12-08T17:32:00+00:00',
    LastProcessingResult: 'OK',
    State: 'Deleting',
    StateTransitionReason: 'User action',
    DestinationConfig: {
      OnFailure: {},
    },
    MaximumRecordAgeInSeconds: -1,
    BisectBatchOnFunctionError: false,
    MaximumRetryAttempts: -1,
  },
};

const _lambdaConstructor = jest.fn();
const _listEventSourceMappings = jest.fn(
  async () => _lambdaResponses.listEventSourceMappings
);
const _updateEventSourceMapping = jest.fn(
  async () => _lambdaResponses.updateEventSourceMapping
);
const _createEventSourceMapping = jest.fn(
  async () => _lambdaResponses.createEventSourceMapping
);
const _deleteEventSourceMapping = jest.fn(
  async () => _lambdaResponses.deleteEventSourceMapping
);

class Lambda {
  constructor(params) {
    _lambdaConstructor(params);
  }
  listEventSourceMappings(params) {
    return { promise: _listEventSourceMappings.bind(this, params) };
  }
  updateEventSourceMapping(params) {
    return { promise: _updateEventSourceMapping.bind(this, params) };
  }
  createEventSourceMapping(params) {
    return { promise: _createEventSourceMapping.bind(this, params) };
  }
  deleteEventSourceMapping(params) {
    return { promise: _deleteEventSourceMapping.bind(this, params) };
  }
}

Lambda._lambdaConstructor = _lambdaConstructor;
Lambda._listEventSourceMappings = _listEventSourceMappings;
Lambda._updateEventSourceMapping = _updateEventSourceMapping;
Lambda._createEventSourceMapping = _createEventSourceMapping;
Lambda._deleteEventSourceMapping = _deleteEventSourceMapping;
Lambda._lambdaResponses = _lambdaResponses;

const _eventsConstructor = jest.fn();
const _eventsResponses = {
  listRuleNamesByTarget: {
    RuleNames: ['fn-dev-RefreshEventsRuleSchedule-1SI6FNZ1LC3NI'],
  },
};
const _enableRule = jest.fn(async () => _eventsResponses.enableRule);
const _disableRule = jest.fn(async () => _eventsResponses.disableRule);
const _listRuleNamesByTarget = jest.fn(
  async () => _eventsResponses.listRuleNamesByTarget
);

class CloudWatchEvents {
  constructor(params) {
    _eventsConstructor(params);
  }
  enableRule(params) {
    return { promise: _enableRule.bind(this, params) };
  }
  disableRule(params) {
    return { promise: _disableRule.bind(this, params) };
  }
  listRuleNamesByTarget(params) {
    return { promise: _listRuleNamesByTarget.bind(this, params) };
  }
}

CloudWatchEvents._eventsResponses = _eventsResponses;
CloudWatchEvents._enableRule = _enableRule;
CloudWatchEvents._disableRule = _disableRule;
CloudWatchEvents._listRuleNamesByTarget = _listRuleNamesByTarget;

const _cwConstructor = jest.fn();
const _cwResponses = {
  describeAlarms: {
    MetricAlarms: [
      {
        AlarmName: 'mock service - dev - lambda1 - B',
      },
      {
        AlarmName: 'mock service - dev - lambda2 - B',
      },
      {
        AlarmName: 'mock service - dev - lambda3 - B',
      },
      {
        AlarmName: 'mock service - dev - lambda1 - A',
      },
      {
        AlarmName: 'mock service - dev - lambda2 - A',
      },
      {
        AlarmName: 'mock service - dev - lambda3 - A',
      },
    ],
  },
  getMetricData: {
    MetricDataResults: [
      {
        Id: 'iteratorage',
        Label: 'IteratorAge',
        Timestamps: [
          new Date('2021-06-02T15:23:00.000Z'),
          new Date('2021-06-02T15:22:00.000Z'),
          new Date('2021-06-02T15:21:00.000Z'),
          new Date('2021-06-02T15:20:00.000Z'),
          new Date('2021-06-02T15:19:00.000Z'),
        ],
        Values: [5193039, 5121276, 5065201, 5010924, 4955845],
        StatusCode: 'Complete',
        Messages: [],
      },
      {
        Id: 'errors',
        Label: 'Errors',
        Timestamps: [
          new Date('2021-06-02T15:23:00.000Z'),
          new Date('2021-06-02T15:22:00.000Z'),
          new Date('2021-06-02T15:21:00.000Z'),
          new Date('2021-06-02T15:20:00.000Z'),
          new Date('2021-06-02T15:19:00.000Z'),
        ],
        Values: [4, 4, 4, 4, 4],
        StatusCode: 'Complete',
        Messages: [],
      },
      {
        Id: 'concurrentexecutions',
        Label: 'ConcurrentExecutions',
        Timestamps: [
          new Date('2021-06-02T15:23:00.000Z'),
          new Date('2021-06-02T15:22:00.000Z'),
          new Date('2021-06-02T15:21:00.000Z'),
          new Date('2021-06-02T15:20:00.000Z'),
          new Date('2021-06-02T15:19:00.000Z'),
        ],
        Values: [1, 1, 1, 1, 1],
        StatusCode: 'Complete',
        Messages: [],
      },
      {
        Id: 'invocations',
        Label: 'Invocations',
        Timestamps: [
          new Date('2021-06-02T15:23:00.000Z'),
          new Date('2021-06-02T15:22:00.000Z'),
          new Date('2021-06-02T15:21:00.000Z'),
          new Date('2021-06-02T15:20:00.000Z'),
          new Date('2021-06-02T15:19:00.000Z'),
        ],
        Values: [4, 4, 4, 4, 4],
        StatusCode: 'Complete',
        Messages: [],
      },
      {
        Id: 'duration',
        Label: 'Duration',
        Timestamps: [
          new Date('2021-06-02T15:23:00.000Z'),
          new Date('2021-06-02T15:22:00.000Z'),
          new Date('2021-06-02T15:21:00.000Z'),
          new Date('2021-06-02T15:20:00.000Z'),
          new Date('2021-06-02T15:19:00.000Z'),
        ],
        Values: [10.4375, 12.7475, 10.685, 10.4425, 11.1325],
        StatusCode: 'Complete',
        Messages: [],
      },
      {
        Id: 'throttles',
        Label: 'Throttles',
        Timestamps: [
          new Date('2021-06-02T15:23:00.000Z'),
          new Date('2021-06-02T15:22:00.000Z'),
          new Date('2021-06-02T15:21:00.000Z'),
          new Date('2021-06-02T15:20:00.000Z'),
          new Date('2021-06-02T15:19:00.000Z'),
        ],
        Values: [0, 0, 0, 0, 0],
        StatusCode: 'Complete',
        Messages: [],
      },
    ],
  },
};
const _enableAlarmActions = jest.fn();
const _disableAlarmActions = jest.fn();
const _describeAlarms = jest.fn(async () => _cwResponses.describeAlarms);
const _getMetricData = jest.fn(async () => _cwResponses.getMetricData);

class CloudWatch {
  constructor(params) {
    _cwConstructor(params);
  }
  describeAlarms(params) {
    return { promise: _describeAlarms.bind(this, params) };
  }
  disableAlarmActions(params) {
    return { promise: _disableAlarmActions.bind(this, params) };
  }
  enableAlarmActions(params) {
    return { promise: _enableAlarmActions.bind(this, params) };
  }
  getMetricData(params) {
    return { promise: _getMetricData.bind(this, params) };
  }
}
CloudWatch._cwResponses = _cwResponses;
CloudWatch._enableAlarmActions = _enableAlarmActions;
CloudWatch._disableAlarmActions = _disableAlarmActions;
CloudWatch._describeAlarms = _describeAlarms;
CloudWatch._getMetricData = _getMetricData;

const _kinesisResponses = {
  registerStreamConsumer: {
    Consumer: {
      ConsumerName: 'con1',
      ConsumerARN:
        'arn:aws:kinesis:eu-central-1:123456789:stream/my-stream-dev/consumer/con1:1607511009',
      ConsumerStatus: 'CREATING',
      ConsumerCreationTimestamp: '2020-12-09T10:50:09+00:00',
    },
  },
  deregisterStreamConsumer: {},
  describeStreamConsumer: {
    ConsumerDescription: {
      ConsumerName: 'con1',
      ConsumerARN:
        'arn:aws:kinesis:eu-central-1:123456789:stream/my-stream-dev/consumer/con1:1607511009',
      ConsumerStatus: 'ACTIVE',
      ConsumerCreationTimestamp: '2020-12-09T10:50:09+00:00',
      StreamARN: 'arn:aws:kinesis:eu-central-1:123456789:stream/my-stream-dev',
    },
  },
};
const _kinesisConstructor = jest.fn();
const _registerStreamConsumer = jest.fn(
  async () => _kinesisResponses.registerStreamConsumer
);
const _deregisterStreamConsumer = jest.fn(
  async () => _kinesisResponses.deregisterStreamConsumer
);
const _describeStreamConsumer = jest.fn(
  async () => _kinesisResponses.describeStreamConsumer
);

class Kinesis {
  constructor(params) {
    _kinesisConstructor(params);
  }
  registerStreamConsumer(params) {
    return { promise: _registerStreamConsumer.bind(this, params) };
  }
  deregisterStreamConsumer(params) {
    return { promise: _deregisterStreamConsumer.bind(this, params) };
  }
  describeStreamConsumer(params) {
    return { promise: _describeStreamConsumer.bind(this, params) };
  }
}

Kinesis._kinesisConstructor = _kinesisConstructor;
Kinesis._registerStreamConsumer = _registerStreamConsumer;
Kinesis._deregisterStreamConsumer = _deregisterStreamConsumer;
Kinesis._describeStreamConsumer = _describeStreamConsumer;
Kinesis._kinesisResponses = _kinesisResponses;

module.exports = {
  ApplicationAutoScaling,
  CloudWatchEvents,
  CloudWatch,
  DynamoDB,
  Kinesis,
  Lambda,
  RDS,
  SNS,
  SQS,
};
